### PR TITLE
fix: Add batching strategy to segment analytics tracking so that events are sent to segment in batches instead of individual

### PR DESCRIPTION
**Pull request by Rhitottam**
> Fixes https://github.com/appsmithorg/appsmith/issues/18829

> From Analytics 2.0 onwards, batching to deliver of multiple events with fewer API calls for faster performance
[https://segment.com/blog/analytics-js-2/](https://segment.com/blog/analytics-js-2/)
[https://segment.com/docs/connections/sources/catalog/libraries/website/javascript/#batching](https://segment.com/docs/connections/sources/catalog/libraries/website/javascript/#batching)

## Description

> Add batching strategy to segment analytics while loading the script inside AnalyticsUtil
> [https://segment.com/docs/connections/sources/catalog/libraries/website/javascript/#batching](https://segment.com/docs/connections/sources/catalog/libraries/website/javascript/#batching)

Why are we doing this, i.e, exact issue with the performance?
 - Whenever any operation is done inside Appsmith, multiple events are being tracked and sent to Segment for tracking. Due to this a lot apis calls are made for each event being tracked using the segment api. This can cause performance issues since for every event being tracked, segment runs a microtask for updating cookies  which only start running when the api is being called. Through batching these events in groups of 100, all microtasks for those events in a group can run without any waiting time in between, thus improving performance slightly.
Screenshot 2023-01-26 at 10 39 09 AM
Screenshot 2023-01-26 at 10 50 04 AM

What are we doing exactly?
- Adding a batching strategy to the Segment config so that events are tracked and delivered in batches of 100 instead of individually delivering them to Segment

Which areas of the application will be effected?
- This will affect the timing of the events being tracked by Segment. Because of this events in a batch of 100 will be received at the same time rather than then sequentially

Will the users see any effects?
- User will see a slight to moderate improvement in performance while doing operations like loading a page in Edit mode or reloading an application in Edit mode

 Fixes https://github.com/appsmithorg/appsmith/issues/18829

## Type of change

> Please delete options that are not relevant.

- Bug fix (non-breaking change which fixes an issue)
- Chore (housekeeping or task changes that don't impact user perception)
- This change requires a documentation update

- Manual

### Dev activity
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

### QA activity:
- [ ] Test plan has been approved by relevant developers
- [ ] Test plan has been peer reviewed by QA
- [ ] Cypress test cases have been added and approved by either SDET or manual QA
- [ ] Organized project review call with relevant stakeholders after Round 1/2 of QA
- [ ] Added Test Plan Approved label after reveiwing all Cypress test

### CODE REVIEW

**Review by sbalaji1192 — APPROVED** [review]

**Review by riodeuno — COMMENTED** [review]
@Rhitottam This looks good to me. Please add comments in the code to describe what each of the options do, particularly, `strategy`, `size` and `timeout`. Also a link to the API docs would be helpful. This way, I would quickly understand what these are in a year's time 😅

**Comment by riodeuno** [comment]
@Rhitottam could you also mention in the PR description the answer to the following questions:
1. Why are we doing this, i.e, exact issue with the performance?
2. What are we doing exactly?
3. Which areas of the application will be effected?
4. Will the users see any effects?

**Comment by Rhitottam** [comment]
Hi @riodeuno , I have the code comments and added an explanation for the changes done.

**Review by riodeuno — DISMISSED** [review]

**Comment by riodeuno** [comment]
@Rhitottam This is perfect. Thanks!. Looks good to me. I've approved the PR.

File: app/client/src/utils/AnalyticsUtil.tsx [hunk 1]
```diff
@@ -385,7 +385,20 @@ class AnalyticsUtil {
...
-          analytics.load(key);
+          // Ref: https://segment.com/docs/connections/sources/catalog/libraries/website/javascript/#batching
+          analytics.load(key, {
+            integrations: {
+              "Segment.io": {
+                deliveryStrategy: {
+                  strategy: "batching", // The delivery strategy used for sending events to Segment
+                  config: {
+                    size: 100, // The batch size is the threshold that forces all batched events to be sent once it’s reached.
+                    timeout: 5000, // The number of milliseconds that forces all events queued for batching to be sent, regardless of the batch size, once it’s reached
```
Comment (SatishGandham) [inline app/client/src/utils/AnalyticsUtil.tsx:396]: Change the batching to one second.

**Review by SatishGandham — CHANGES_REQUESTED** [review]

File: app/client/src/utils/AnalyticsUtil.tsx (hunk 1 above)
Comment (Rhitottam) [reply app/client/src/utils/AnalyticsUtil.tsx]: Done!!

**Review by eco-monk — APPROVED** [review]
Looks good to me.

**Review by SatishGandham — APPROVED** [review]

**Pull request merged**